### DISCUSSION TITLE
Set Slider Value from Machine Data

### DISCUSSION THREAD

**Discussion by supersebbo** [artisan-roaster-scope/artisan#1442] [General]
Hi folks,

I'm building a websockets integration for a home-build roaster.  I have got most things working but one challenge I can't fathom is how to set slider values based on data from the machine.  This is mostly for initialization as I notice the sliders currently remember their last setting in Artisan rather than the actual value from the machine. 

What I want to achieve is a two-way control, so the slider value is updated from the machine, and changes to the slider are sent to the machine.  I can do the latter quite easily:

`send({{"command": "burner", "value":"{}"}})`

This is working.  But how to read the value from the machine and update the slider in Artisan?

Assume the current burner value is in the `getData` response at node `["data"]["burner"]`.

There is an example for button state setting in the docs, but not slider state setting. 
`read({{"command": "getRoasterState"}});button(_["data"]["state"] == "ready")`

Thanks!

**Comment by MAKOMO** [comment]
Artisan has the concept of [Event Quantification](https://artisan-roasterscope.blogspot.com/2014/04/event-quantifiers.html). The idea is that a stream of data gathered from regular sampling can be observed by such an event quantifier and on significant changes corresponding events are generated. So if the data stream represents burner values in the range from 0-100, such a quantifier can insert a gas change event "gas 60%" on a drop of the burner from 80 to 60. If that event type has a slider associated it will be moved to the new value, 60 in that case.

**Reply by supersebbo** [reply]
Hi, thanks for the suggestion.  I think the problem here is that this slider<->machine state inconsistency often occurs at the start of a roast.  During the previous roast I will have set the status of the heater (which is indeed 0-100) to an arbitrary value (say 75%) but another event automation then shuts down the heater to 0 when the 'Drop' button is pressed.  So the slider is still set at 75% but the actual heater value is now 0%. 

On the start of the next roast the machine sends it's actual value (0%) but the slider is still set at 75%, remembered from the previous session. As there is no 'change' in the heater value at this point, there is nothing to trigger an event on. 

Perhaps it's a feature request, but it would be great to be able to set slider values similar to button values:
This works for buttons:
`read({{"command": "getRoasterState"}});button(_["data"]["state"] == "ready")`

Suggested syntax for sliders would be:
`read({{"command": "getRoasterState"}});slider(_["data"]["state"])`

Or something to that effect

**Reply by MAKOMO** [reply]
I don't know your exact setup, but an event slider is always representing the last event value. If your "other event automation" shuts down the heater to 0 it should also add 0% burner event. This can be achieved via a simple event button which might itself be hidden if needed.

<img width="1045" alt="Screenshot 2024-02-10 at 12 30 04" src="https://github.com/artisan-roaster-scope/artisan/assets/13909011/7ab7696a-7a74-46de-8d64-9400b75056be">
